pytket/extensions/cutensornet/mps/mps_mpo.py

```
@@ -0,0 +1,572 @@
+            pos: int, left_tensor: Optional[Tensor], right_tensor: Optional[Tensor]
+        ) -> float:
+            """Update the tensor at ``pos`` of the variational MPS using ``left_tensor``
+            (and ``right_tensor``) which is meant to contain the contraction of all
+            the left (and right) columns of the MPS-MPO-vMPS* network from ``pos``.
+            Contract these with the MPS-MPO column at ``pos``.
+            Return the current fidelity of this sweep.
+            """
+            interleaved_rep = [
+                # The tensor of the MPS
+                self.tensors[pos].data,
+                self.tensors[pos].bonds,
+            ]
+            # The MPO tensors at position ``pos``
+            for mpo_tensor in self._mpo[pos]:
+                interleaved_rep.append(mpo_tensor.data)
+                interleaved_rep.append(mpo_tensor.bonds)
+            # The output bond of the last tensor must connect to the physical
+            # bond of the corresponding ``self._aux_mps`` tensor
+            interleaved_rep[-1][-1] = -self._aux_mps.get_physical_bond(pos)
+
+            if left_tensor is not None:
+                interleaved_rep.append(left_tensor.data)
+                interleaved_rep.append(left_tensor.bonds)
+            if right_tensor is not None:
+                interleaved_rep.append(right_tensor.data)
+                interleaved_rep.append(right_tensor.bonds)
+
+            # Append the bond IDs of the resulting tensor
+            F_bonds = [-b for b in self._aux_mps.tensors[pos].bonds]
+            interleaved_rep.append(F_bonds)
+
+            # Contract and store tensor
+            F = Tensor(cq.contract(*interleaved_rep), self._aux_mps.tensors[pos].bonds)
+
+            # Get the fidelity
+            optim_fidelity = complex(
+                cq.contract(F.data.conj(), F.bonds, F.data, F.bonds, [])
+            )
+            assert np.isclose(optim_fidelity.imag, 0.0, atol=self._atol)

```

Since this (and other) assert is not guaranteed to be executed at runtime, it is worth thinking if one should raise an error here if assertion is not satisfied.

---

My use of `assert` statements restricts to sanity checks. By sanity checks I mean statements that should _always_ be true, no matter the user's input. They are just there to check that the code is doing what it's meant to do; they are meant just for the "debug" mode.